File: lib/expect-config.js
"use strict";

var _ = require('underscore')._;

module.exports = function expectConfig(config, debug) {
    var lines = [];

    if (!config.hosts[config.general.main]) {
        throw new Error('Configuration does not contain information about host "' + config.general.main + '"');
    }

    if (!debug) {
        lines.push('log_user 0');
    }

    lines.push('set timeout 30');

    if (debug) {
        lines.push('spawn ssh -v -F ' + config.sshConfig + ' ' + config.general.main);
    } else {
        lines.push('spawn ssh -F ' + config.sshConfig + ' ' + config.general.main);
    }

    lines.push('expect {');
    _.each(config.hosts, function (c, name) {
        if (c.user && c.password) {
            lines.push('  # ' + name);
            lines.push('  "' + c.user + '@' + c.addr + '" {');
            lines.push('    send "' + c.password + '\\n";');
            lines.push('    exp_continue;');
            lines.push('  }');

            if (name === config.general.main) {
                lines.push('  # ' + name + ' as only host');
                lines.push('  "Password:" {');
                lines.push('    send "' + c.password + '\\n";');
                lines.push('    exp_continue;');
                lines.push('  }');
            }
        }
    });

    if (config.hosts[config.general.main].prompt) {
        lines.push('  -re "' + config.hosts[config.general.main].prompt + '" {');
    } else {
        lines.push('  -re "(%|\\\\$|#|>)\\\\s*$" {');
    }
    lines.push('    send_user "\\nThe login sequence seems to have worked.\\n\\n";');
    if (config.forwards) {
        lines.push('    send_user "The following forwardings have been set up for you:\\n\\n";');
        _.each(config.forwards, function (fs, descr) {
            lines.push('    send_user "' + descr + ':\\n";');
            _.each(fs, function (to, from) {
                lines.push('    send_user "   ' + from + ' -> ' + to + '\\n";');
            });
            lines.push('    send_user "\\n";');
        });
    }
    lines.push('    send "\\r";');
    lines.push('    interact;');
    lines.push('  }');
    lines.push('  "Permission denied" {');
    lines.push('    send_user "Permission denied, failed to set up tunneling.\n";');
    lines.push('    exit 2;');
    lines.push('  }');
    lines.push('  timeout {');
    lines.push('    send_user "Unknown error, failed to set up tunneling.\n";');
    lines.push('    exit 2;');
    lines.push('  }');
    lines.push('}');

    // Propagate exit code
    lines.push('catch wait reason');
    lines.push('exit [lindex $reason 3]');

    return lines.join('\n') + '\n';
};

File: expect-config.go
package main

import (
	"fmt"
	"log"
	"strings"

	"github.com/calmh/mole/configuration"
	"github.com/calmh/mole/tmpfileset"
)

func expectConfig(cfg *configuration.Config, fs *tmpfileset.FileSet) {
	if cfg.General.Main == "" {
		log.Fatal("Cannot generate expect script for empty destination host")
	}

	_, ok := cfg.HostsMap[cfg.General.Main]
	if !ok {
		log.Fatalf("Cannot generate expect script for non-existent host %q", cfg.General.Main)
	}

	var lines []string
	nl := func(line string) {
		lines = append(lines, line)
	}

	nl("set timeout 30")
	if globalOpts.Debug {
		nl("spawn ssh -v -F {ssh-config} " + cfg.General.Main)
	} else {
		nl("spawn ssh -F {ssh-config} " + cfg.General.Main)
	}

	nl("expect {")
	for _, host := range cfg.Hosts {
		if host.User != "" && host.Pass != "" {
			nl("  # " + host.Name)
			nl(fmt.Sprintf(`  "%s@%s" {`, host.User, host.Addr))
			nl(fmt.Sprintf(`    send %q; send "\n";`, host.Pass))
			nl("    exp_continue;")
			nl("  }")

			if host.Name == cfg.General.Main {
				nl("  # " + host.Name + " (as main)")
				nl(`  "Password:" {`)
				nl(fmt.Sprintf(`    send %q; send "\n";`, host.Pass))
				nl("    exp_continue;")
				nl("  }")
			}
		} else {
			nl("  # " + host.Name + " does not need password authentication")
		}
		nl("")
	}

	nl("  # prompt")
	nl(fmt.Sprintf("  -re %q {", cfg.Hosts[cfg.HostsMap[cfg.General.Main]].Prompt))
	nl(`    send_user "\nThe login sequence seems to have worked.\n\n";`)
	nl(`    send "\r";`)

	for _, fwd := range cfg.Forwards {
		nl(`    send_user "` + fwd.Name + `:\n";`)
		for _, line := range fwd.Lines {
			if line.Repeat > 0 {
				nl(fmt.Sprintf("    send_user \"    %s %d-%d -> %s %d-%d\\n\";", line.SrcIP, line.SrcPort, line.SrcPort+line.Repeat, line.DstIP, line.DstPort, line.DstPort+line.Repeat))
			} else {
				nl(fmt.Sprintf("    send_user \"    %s %d -> %s %d\\n\";", line.SrcIP, line.SrcPort, line.DstIP, line.DstPort))
			}
		}
		nl(`    send_user "\n";`)
	}

	nl("    interact;")
	nl("  }")
	nl("")

	nl(`  "Permission denied" {`)
	nl(`    send_user "\nPermission denied, failed to set up tunneling.\n\n";`)
	nl(`    exit 2;`)
	nl("  }")
	nl("")

	nl(`  timeout {`)
	nl(`    send_user "\nUnknown error, failed to set up tunneling.\n\n";`)
	nl(`    exit 2;`)
	nl("  }")

	nl("}")

	nl("catch wait reason;")
	nl("exit [lindex $reason 3];")

	content := []byte(strings.Join(lines, "\n") + "\n")
	fs.Add("expect-config", content)
}
